internal/state/graph/gateway_listener.go

```diff
@@ -325,27 +334,30 @@ func createHTTPSListenerValidator(gwNsName string) listenerValidator {
 }
 
 func createHostnameConflictResolver() listenerConflictResolver {
```

Comment:
we need to handle conflicts when listeners for _different_ protocols try to use the same port.

For example:
```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: gateway
  labels:
    domain: k8s-gateway.nginx.org
spec:
  gatewayClassName: nginx
  listeners:
  - name: http
    port: 80
    protocol: HTTP
  - name: https
    port: 80
    protocol: HTTPS
    tls:
      mode: Terminate
      certificateRefs:
      - kind: Secret
        name: cafe-secret
        namespace: default
```

Right now, it passes our validation and results into erronous NGINX config:
```nginx
server {
    listen 80 default_server;

    default_type text/html;
    return 404;
}
server {
    listen 80 ssl default_server;

    ssl_reject_handshake on;
}

server {
    listen 80 ssl;
    ssl_certificate /etc/nginx/secrets/default_cafe-secret;
    ssl_certificate_key /etc/nginx/secrets/default_cafe-secret;

    if ($ssl_server_name != $host) {
        return 421;
    }

    server_name ~^;


    location / {
        return 404 "";

    }
}
```

Reply:
Yep, good catch. I need to change the primary map key to port.

Reply:
Should we  handle port conflicts separately from hostname conflicts?

Here is another problematic Gateway:
```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: gateway
  labels:
    domain: k8s-gateway.nginx.org
spec:
  gatewayClassName: nginx
  listeners:
  - name: http
    port: 443
    protocol: HTTP
    hostname: example.com
  - name: https
    port: 443
    protocol: HTTPS
    tls:
      mode: Terminate
      certificateRefs:
      - kind: Secret
        name: cafe-secret
        namespace: default
````        

```nginx
server {
    listen 443 default_server;

    default_type text/html;
    return 404;
}
server {
    listen 443 ssl default_server;

    ssl_reject_handshake on;
}

server {
    listen 443 ssl;
    ssl_certificate /etc/nginx/secrets/default_cafe-secret;
    ssl_certificate_key /etc/nginx/secrets/default_cafe-secret;

    if ($ssl_server_name != $host) {
        return 421;
    }

    server_name ~^;


    location / {
        return 404 "";

    }

}
```

the issue: we can't ask NGINX to handle both HTTP and HTTPS on the same port.

Reply:
Spoke offline. In this case, we should mark each listener with `Accepted:False` and add a new Condition Reason along the lines of "listener conflict; cannot have an http and https listener on the same port"

I will also ask if this config is valid in the gateway api slack.

Reply:
Fixed. Since the webhook checks for hostname conflicts, I removed the hostname conflict resolver from this package. I also added a test for hostname conflicts to the change processor's webhook assumptions test.